### ISSUE TITLE
Tess4j on OpenCL / Multithread

### ISSUE THREAD

**Issue by sonik340** [nguyenq/tess4j#24]
Hello,
I have seen on the Internet that it is possible to use tesseract with OpenCL.
Could tess4j support it ? It would be a tremendous feature.
Vulkan ? But it may be a bit early :-)

**Comment by 4F2E4A2E** [comment]
Tess4J uses jni to access the tesseract native lib, because of that i imagine to be very difficult to get it running via OpenCL, but nothing is impossible.

So why should we support this? It would not be a feature of our Project, but instead a Project itself.

By the way, i guess for the most use cases and "Java Thread Skills" you won't really need it and i don't think Enterprise Companys like NSA, etc. is using Tess4J for OCR, or are they? ^^

Vulkan!?

**Comment by sonik340** [comment]
The company I work for will use tess4j to read thousands of document a day.
They are interested in performance but powerful servers could cost an arm
:-) That is why I wander about way to fasten tess4j.
I have found some who built tesseract to use opencl so, knowing that tess4j
is "only" an interface I wonder if the work to get an opencl-tess4j was
possible.

Vulkan : I just heard about this. As far as I I understand it is a
framework that gives direct access to the graphic card. Do not mind about
that.

**Comment by 4F2E4A2E** [comment]
Hey @sonik340 are you still on to this?

**Comment by sonik340** [comment]
Hi,
This is not a priority for now. As far as I know, my company will not pay
for special servers.
But of course, any improvement would be welcomed. Indeed, I never made it
to make Tesseract work on multiple threads.

2016-03-01 23:31 GMT+01:00 4F2E4A2E <notifications@github.com>:

> Hey @sonik340 <https://github.com/sonik340> are you still on to this?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nguyenq/tess4j/issues/24#issuecomment-190936633>.
>



-- 
SMORDOWSKI Romain
Adresse: 129 rue Jules Guesde, 59170 Croix
Courriel: romain.smordowski@gmail.com
Tél. portable: 06 80 47 87 36
Tél. fixe: 09 54 68 06 90

**Comment by 4F2E4A2E** [comment]
Tesseract guys has added support for [OpenCL](https://github.com/tesseract-ocr/tesseract/releases), j have no clue about it, but if this topic (thread conversion) interests me a lot, if you are up to it we could try to get this done. What do say?

**Comment by sonik340** [comment]
Sadly, I do not have a lot of spare time. If you have precise task I can
help let me know. I am not Java expert but I would like too.

**Comment by 4F2E4A2E** [comment]
Do you have any OpenCL Example i can build on? In Java!?

**Comment by sonik340** [comment]
Sadly no. But there is an AMD api that can help do the job. Or may be
Tesseract already use OpenCL and we just have to JNI on it.

**Comment by sonik340** [comment]
Here is a very interesting link for java openCL :
http://developer.amd.com/tools-and-sdks/opencl-zone/aparapi/

I found out on Google that we would just have to enable openCL on tesseract
but I do not precisely found how to (seems to be a shell option)

**Comment by 4F2E4A2E** [comment]
Found some information on how to do it natively: http://www.sk-spell.sk.cx/tesseract-meets-the-opencl-first-test

**Comment by sonik340** [comment]
Fine !
I do not know how you plugged java to Tesseract. Can you give me
information on how you did it (technics used). I think it is JNI.
I so I will try to get to know more about it. A good example would help.

**Comment by 4F2E4A2E** [comment]
Well sure, but for that you just have to take a look to our source code. I still think that we can solve this using Threads in Java. 

Is the goal merely parallel processing?

**Comment by sonik340** [comment]
I already tried parallel processing but I got errors from Tesseract, I
think because two instances used the same files.

**Comment by 4F2E4A2E** [comment]
Well this depend on so many things, real threads in java are not that easy, but trying to convert the same image many times maybe not the best aproach.

http://stackoverflow.com/a/6983094/543426

I will try to solve this with java extracting txt from 100 images in parallel, is that ok for you?

**Comment by sonik340** [comment]
I tried to launch one tesseract instance per image (on different images).
The problems seemed to come from a needed tesseract reference file.

**Comment by nguyenq** [comment]
There is a nice example of executing `doOCR` in parallel at https://sourceforge.net/p/tess4j/discussion/1202293/thread/4562eccb/

**Comment by sonik340** [comment]
So, we just need to upgrade Tesseract inside tess4j ?

**Comment by 4F2E4A2E** [comment]
I already have an working example for you. I will deliver it soon.

**Comment by 4F2E4A2E** [comment]
There you go: multi threaded conversion:
```
-------------------------------------------------------
 T E S T S
-------------------------------------------------------
Running net.sourceforge.tess4j.TestExtractTextFromImageMultiThreaded
Starting conversion with '4' Threads.
Extracted: \tess4j-multi-threads\[...]batch-conversion\eurotext_deskew.png.txt
Extracted: \tess4j-multi-threads\[...]batch-conversion\eurotext_03.png.txt
Extracted: \tess4j-multi-threads\[...]batch-conversion\eurotext_02.png.txt
Extracted: \tess4j-multi-threads\[...]batch-conversion\eurotext_deskew_04.png.txt
Extracted: \tess4j-multi-threads\[...]batch-conversion\eurotext_unlv.png.txt
Extracted: \tess4j-multi-threads\[...]batch-conversion\eurotext_deskew_03.png.txt
Extracted: \tess4j-multi-threads\[...]batch-conversion\eurotext_unlv_04.png.txt
Extracted: \tess4j-multi-threads\[...]batch-conversion\eurotext_unlv_03.png.txt
Extracted: \tess4j-multi-threads\[...]batch-conversion\eurotext.png.txt
Extracted: \tess4j-multi-threads\[...]batch-conversion\eurotext_deskew_02.png.txt
Extracted: \tess4j-multi-threads\[...]batch-conversion\eurotext_unlv_02.png.txt
Extracted: \tess4j-multi-threads\[...]batch-conversion\eurotext_04.png.txt
Finished all threads
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 7.65 sec
```
```
Running net.sourceforge.tess4j.TestExtractTextFromImageSingleThreaded
Starting conversion with '1' Threads.
Extracted: \tess4j-multi-threads\[...]batch-conversion\eurotext_unlv.png.txt
Extracted: \tess4j-multi-threads\[...]batch-conversion\eurotext_deskew_04.png.txt
Extracted: \tess4j-multi-threads\[...]batch-conversion\eurotext_deskew_03.png.txt
Extracted: \tess4j-multi-threads\[...]batch-conversion\eurotext_unlv_04.png.txt
Extracted: \tess4j-multi-threads\[...]batch-conversion\eurotext_unlv_03.png.txt
Extracted: \tess4j-multi-threads\[...]batch-conversion\eurotext_unlv_02.png.txt
Extracted: \tess4j-multi-threads\[...]batch-conversion\eurotext_deskew.png.txt
Extracted: \tess4j-multi-threads\[...]batch-conversion\eurotext_04.png.txt
Extracted: \tess4j-multi-threads\[...]batch-conversion\eurotext_deskew_02.png.txt
Extracted: \tess4j-multi-threads\[...]batch-conversion\eurotext_03.png.txt
Extracted: \tess4j-multi-threads\[...]batch-conversion\eurotext_02.png.txt
Extracted: \tess4j-multi-threads\[...]batch-conversion\eurotext.png.txt
Finished all threads
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 10.56 sec
```
Almost half of the time needed, please consider reading [Amdah's Law](http://tutorials.jenkov.com/java-concurrency/amdahls-law.html) and [Vogella's Tutorial on Java-Threads](http://www.vogella.com/tutorials/JavaConcurrency/article.html#concurrency_overview)
[tess4j-multi-threads.zip](https://github.com/nguyenq/tess4j/files/161707/tess4j-multi-threads.zip)

**Comment by 4F2E4A2E** [comment]
@sonik340 any status update for us would be appreciated in order to close this issue.
